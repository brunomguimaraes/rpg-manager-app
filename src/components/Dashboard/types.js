// @flow

import type { ThunkAction } from "~/types/redux";
import {
  FETCH_DASHBOARD_DATA_PENDING,
  FETCH_DASHBOARD_DATA_SUCCESS,
  FETCH_DASHBOARD_DATA_FAILURE
} from "./constants";
import type { Character } from "~/components/CharacterManager";

export type CharactersByGame = {
  gameId: number,
  gameName: string,
  characters: Array<Character>
};

export type FetchDashboardDataActionTypes =
  | { +type: typeof FETCH_DASHBOARD_DATA_PENDING }
  | { +type: typeof FETCH_DASHBOARD_DATA_SUCCESS }
  | { +type: typeof FETCH_DASHBOARD_DATA_FAILURE };

export type FetchDashboardDataAction = {
  payload: Array<CharactersByGame>
} & FetchDashboardDataActionTypes;

export type Action = FetchDashboardDataAction;

export type State = {
  error: boolean,
  loading: boolean,
  charactersByGame: Array<CharactersByGame>
};

export type GetDashboardData = () => ThunkAction;

export type Props = State & {
  getDashboardData: GetDashboardData
};
