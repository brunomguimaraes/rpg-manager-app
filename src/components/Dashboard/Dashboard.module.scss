@import "~styles/variables";

.dashboard {
  width: 100%;
  margin: 0;
  padding: 0;
  display: block;
  overflow: hidden;
  padding: $spacer * 2;
  list-style-type: none;
  box-sizing: border-box;
  @media screen and (max-width: $break-small) {
    padding: $spacer;
  }
  & > li {
    padding: $spacer * 2;
    @media screen and (max-width: $break-small) {
      padding: $spacer;
    }
  }
}
