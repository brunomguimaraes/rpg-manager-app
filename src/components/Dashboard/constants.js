// @flow

const FETCH_DASHBOARD_DATA = "FETCH_DASHBOARD_DATA";
const FETCH_DASHBOARD_DATA_PENDING = `${FETCH_DASHBOARD_DATA}_PENDING`;
const FETCH_DASHBOARD_DATA_SUCCESS = `${FETCH_DASHBOARD_DATA}_SUCCESS`;
const FETCH_DASHBOARD_DATA_FAILURE = `${FETCH_DASHBOARD_DATA}_FAILURE`;

export {
  FETCH_DASHBOARD_DATA,
  FETCH_DASHBOARD_DATA_PENDING,
  FETCH_DASHBOARD_DATA_SUCCESS,
  FETCH_DASHBOARD_DATA_FAILURE
};
