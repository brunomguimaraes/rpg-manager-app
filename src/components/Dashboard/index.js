// @flow

import Loadable from "react-loadable";
import Loading from "~/components/Loading";

export { default as dashboardReducer } from "./reducers";
export { getDashboardData } from "./actions";

export default Loadable({
  loader: () => import("./Dashboard"),
  loading: Loading
});

export type { Action, State } from "./types";
