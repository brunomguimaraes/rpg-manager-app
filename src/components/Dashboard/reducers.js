// @flow

import {
  FETCH_DASHBOARD_DATA_PENDING,
  FETCH_DASHBOARD_DATA_SUCCESS,
  FETCH_DASHBOARD_DATA_FAILURE
} from "./constants";
import type { State, Action } from "./types";

export const initialState: State = {
  charactersByGame: [],
  error: false,
  loading: false
};

function reducer(state: State = initialState, action: Action): State {
  const { type, payload } = action;
  switch (type) {
    case FETCH_DASHBOARD_DATA_PENDING:
      return {
        ...state,
        error: false,
        loading: true
      };
    case FETCH_DASHBOARD_DATA_SUCCESS:
      return {
        ...state,
        error: false,
        loading: false,
        charactersByGame: payload
      };
    case FETCH_DASHBOARD_DATA_FAILURE:
      return {
        ...state,
        error: true,
        loading: false
      };
    default:
      return state;
  }
}

export default reducer;
