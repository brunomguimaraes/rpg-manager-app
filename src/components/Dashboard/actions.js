// @flow

import { createAction } from "redux-actions";
import client from "~/api";
import type { GetDashboardData } from "./types";
import { FETCH_DASHBOARD_DATA } from "./constants";

export const getDashboardData: GetDashboardData = createAction(
  FETCH_DASHBOARD_DATA,
  async () => await client.getCharacter()
);
