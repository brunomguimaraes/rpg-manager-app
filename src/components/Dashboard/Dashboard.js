// @flow

import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { bindActionCreators } from "redux";
import Scrollable from "~/components/Scrollable";
import Game from "./Game";
import { getDashboardData } from "./actions";
import styles from "./Dashboard.module.scss";
import type { Props } from "./types";

export class Dashboard extends Component<Props> {
  componentDidMount() {
    this.props.getDashboardData();
  }

  render() {
    return (
      <section className={styles.container}>
        {this.props.charactersByGame.length ? (
          <Scrollable shouldScroll>
            <ul className={styles.dashboard}>
              {this.props.charactersByGame.map(item => (
                <li key={item.gameId}>
                  <Game {...item} />
                </li>
              ))}
            </ul>
          </Scrollable>
        ) : (
          <Scrollable shouldScroll>
            <div className={styles.welcome}>
              <span>
                Welcome to RPG Manager! <br />
                Here you can manage your rpg campaigns as you never did before.{" "}
                <br />
                Since it is you first time, why don't you start by{" "}
                <Link to="/game">create you first campaign</Link>?
              </span>
            </div>
          </Scrollable>
        )}
      </section>
    );
  }
}

function selector({ dashboard }) {
  return { ...dashboard };
}

function dispatcher(dispatch) {
  return bindActionCreators(
    {
      getDashboardData
    },
    dispatch
  );
}

export default connect(selector, dispatcher)(Dashboard);
