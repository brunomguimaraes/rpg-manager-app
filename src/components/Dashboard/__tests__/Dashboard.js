import React from "react";
import { Dashboard } from "../Dashboard";

const props = {
  charactersByGame: [{ gameId: 1 }, { gameId: 2 }, { gameId: 3 }, { gameId: 4 }]
};

test("Dashboard", () => {
  const spy = jest.fn();
  const component = shallow(<Dashboard {...props} getDashboardData={spy} />);
  expect(component).toMatchSnapshot();
  expect(spy).toHaveBeenCalled();
});
