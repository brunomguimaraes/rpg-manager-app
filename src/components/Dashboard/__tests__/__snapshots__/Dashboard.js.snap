// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard 1`] = `
<Scrollbars
  autoHeight={false}
  autoHeightMax={200}
  autoHeightMin={0}
  autoHide={true}
  autoHideDuration={200}
  autoHideTimeout={1000}
  hideTracksWhenNotNeeded={false}
  renderThumbHorizontal={[Function]}
  renderThumbVertical={[Function]}
  renderTrackHorizontal={[Function]}
  renderTrackVertical={[Function]}
  renderView={[Function]}
  tagName="div"
  thumbMinSize={30}
  universal={false}
>
  <ul
    className="dashboard"
  >
    <li
      key="1"
    >
      <Game
        gameId={1}
      />
    </li>
    <li
      key="2"
    >
      <Game
        gameId={2}
      />
    </li>
    <li
      key="3"
    >
      <Game
        gameId={3}
      />
    </li>
    <li
      key="4"
    >
      <Game
        gameId={4}
      />
    </li>
  </ul>
</Scrollbars>
`;
